src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1969,23 +1969,44 @@ case class SymRef(posn: Position, symbol: String) extends AST(posn) {
 case class If(pos: Position, cond: AST, thenTree: AST, elseTree: AST)
   extends AST(pos, Array(cond, thenTree, elseTree)) {
   override def typecheckThis(ec: EvalContext): BaseType = {
-    if (thenTree.`type` != elseTree.`type`)
-      parseError(s"expected same-type `then' and `else' clause, got `${ thenTree.`type` }' and `${ elseTree.`type` }'")
-    else
-      thenTree.`type`
+    (thenTree.`type`, elseTree.`type`) match {
+      case (thenType, elseType) if thenType == elseType => thenType
+      case (thenType: TNumeric, elseType: TNumeric)     => TNumeric.promoteNumeric(Set(thenType, elseType))
+      case _ =>
+        parseError(s"expected same-type `then' and `else' clause, got `${ thenTree.`type` }' and `${ elseTree.`type` }'")
+    }
   }
 
   def eval(ec: EvalContext): () => Any = {
     val f1 = cond.eval(ec)
     val f2 = thenTree.eval(ec)
     val f3 = elseTree.eval(ec)
+
+    // this dance is necessary to avoid serializing `this` which is not serializable
+    // (and it would be an unnecessary leak anyway)
+    val ifType = `type`
+    val thenTreeType = thenTree.`type`
+    val elseTreeType = elseTree.`type`
+
     () => {
+      def promoteNumeric(targetType: TNumeric, valueType: TNumeric, value: Any): Any = targetType match {
+        case TDouble => valueType.makeDouble(value)
+        case TFloat  => value
+        case TLong   => valueType.asInstanceOf[TIntegral].makeLong(value)
+        case TInt    => value
+      }
+      def coerce(valueType: BaseType, value: Any): Any = valueType match {
+        case `ifType`    => value
+        case t: TNumeric => promoteNumeric(ifType.asInstanceOf[TNumeric], t, value)
+        case _           => value
+      }
+
```

Comment:
I think this need not be so complicated -- we've faced this struggle elsewhere in the AST code, and now have some decent abstractions.  Something like this should work:

```scala
val coerce: (Any) => Any = `type` match { // type here is the result type of the AST node, which is the result of the if/then clause
  case t: TNumeric => (a: Any) => t.conv.to
  case _ => (a: Any) => a
}
```

We actually don't currently have the ability to do this exactly as I wrote it -- TNumeric doesn't have a NumericConversion member, so we'd have to match on Type again.  I think adding a `val conv: NumericConversion[_]` to the abstract `TNumeric` and its children would make this very elegant.  

Any thoughts?  Feel free to push back on this!